src/visualisation/geoviewer.py

```diff
@@ -0,0 +1,414 @@
+        self.log_handler = widget_utils.OutputWidgetHandler()
+        self.logger.addHandler(self.log_handler)
+
+        default_map_layer = ipyleaflet.TileLayer(
+            url="https://{s}.tile.openstreetmap.org/{z}/{x}/{y}.png",
+            base=True,
+            max_zoom=19,
+            min_zoom=4,
+        )
+
+        # Note: entries in layer_dict follow the convention:
+        # ipywidgets_layer = layer_dict[type][name][subtype]["layer"]
+        # Layers of type "maps" only have subtype "map".
+        # The layer_dict overrules the ipyleaflet.Map() attribute .layers
+        self.layer_dict = dict(
+            maps=dict(
+                OpenStreetMap=dict(map=dict(layer=default_map_layer, active=True))
+            ),
+            graphs=dict(),
+        )
+        self.layer_style = style.DEFAULT_LAYER_STYLE
+
+        self.graph_subtypes = [
+            "pgons",
+            "graph",
+            "components",
+            "disconnected_nodes",
+            "poorly_connected_nodes",
+        ]
+
+        # Setting the current view of graph and map as traits. Together with layer_dict
+        # these two determine the state of the widget.
+        self.add_traits(
+            current_graph=traitlets.Unicode().tag(sync=True),
+            current_map=traitlets.Unicode().tag(sync=True),
+        )
+        self.current_graph = ""
+        self.current_map = "Map"  # set to the default map added above
+
+        self.logger.info("Viewer successfully initialised.")
+
+    def set_layer_visibility(
+        self, layer_type: str, layer_name: str, layer_subtype: str, active: bool
+    ) -> None:
+        """Set visiblity for a specific layer
+
+        Set the visibility for layer in
+        `layer_dict[layer_type][layer_name][layer_subtype]`.
+
+        Args:
+            layer_type (str): type of layer (e.g. "maps","graphs")
+            layer_name (str): name of layer
+            layer_subtype (str): subtype of layer (e.g. "map","components")
+            active (bool): whether layer is activate (=visible)
+        """
+        self.layer_dict[layer_type][layer_name][layer_subtype]["active"] = active
+
+    def hidde_all_layers(self) -> None:
+        """ Hide all layers in self.layer_dict."""
+        for layer_type, type_dict in self.layer_dict.items():
+            for layer_name in type_dict:
+                if layer_type == "maps":
+                    self.set_layer_visibility(layer_type, layer_name, "map", False)
+                elif layer_type == "graphs":
+                    self.set_layer_visibility(layer_type, layer_name, "graph", False)
+                    self.set_layer_visibility(layer_type, layer_name, "pgons", False)
+                    self.set_layer_visibility(
+                        layer_type, layer_name, "components", False
+                    )
+        self.layer_update()
+
+    def add_layer(self, layer: Union[dict, ipyleaflet.Layer], name=None) -> None:
```

Comment:
is this function used by anything? I couldn't find an explicit call anywhere.

Reply:
It's not used internally in any widget but it overrides an existing ipyleaflet function. It enables users to add custom (non-graph) layers to the viewer and they will show up as map layers in the UI (good for showing satellite data for example)